library/src/main/kotlin/org/codice/compliance/report/Report.kt

```diff
@@ -0,0 +1,264 @@
+     * Resets the list of the current test's exceptions
+     */
+    fun resetCurrentTestExceptions() {
+        currentTestExceptionMessages = mutableSetOf()
+    }
+
+    /**
+     * Resets the map of the current test's exceptions
+     * This is used for testing purposes only
+     */
+    fun resetExceptionMap() {
+        exceptionMessages.clear()
+        exceptionMessages.putAll(emptyExceptionMap)
+        hasExceptions = false
+    }
+
+    /**
+     * Prints the test exceptions
+     */
+    fun printTestExceptions() {
+        if (reportQuietly) {
+            return
+        }
+        currentTestExceptionMessages.forEach {
+            println(Ansi.ansi().fgMagenta().a(it.message).reset())
+        }
+    }
+
+    /**
+     * Writes the report to a file.
+     */
+    fun writeReport() {
+        // GENERAL is never skipped
+        if (!GENERAL.isStarted()) {
+            GENERAL.start()
+        }
+
+        val file = File(REPORT_FILE)
+        file.printWriter().use { writer ->
+            Section.values().forEach { section ->
+                writer.print("\t".repeat(section.level))
+                writer.print(section.title)
+
+                // the top level section will not have a status except GENERAL and SCHEMA
+                if (section.level == 2 || section == GENERAL || section == SCHEMA) {
+                    printExceptions(section, writer)
+                } else {
+                    writer.println()
+                }
+            }
+
+            writer.println()
+            writer.print("NOTE: A list of MUSTs that are hard to test can be found at: ")
+            writer.print(Ansi.ansi().fgBrightBlue().a(NOT_TESTED_LIST).reset())
+            writer.println()
+        }
+
+        println()
+        print("A full report can be found at: ")
+        print(Ansi.ansi().fgBrightBlue().a(file.absolutePath).reset())
+        println()
+        print("NOTE: A list of MUSTs that are hard to test can be found at: ")
+        print(Ansi.ansi().fgBrightBlue().a(NOT_TESTED_LIST).reset())
+        println()
+    }
+
+    private fun printExceptions(it: Section, writer: PrintWriter) {
+        when {
+            exceptionMessages[it] == null ->
+                writer.print("  ${Ansi.ansi().fgYellow().a("SKIPPED").reset()}")
+            exceptionMessages[it]?.isEmpty() == true ->
+                writer.print("  ${Ansi.ansi().fgGreen().a("SUCCESSFUL").reset()}")
+            else -> {
+                writer.print("  ${Ansi.ansi().fgRed().a("FAILED").reset()}")
+                writer.println()
+                if (reportQuietly) {
+                    return
+                }
+                exceptionMessages[it]?.forEach { exc ->
+                    exc.let {
+                        writer.println(Ansi.ansi().fgMagenta().a(it.message).reset())
+                    }
+                }
+            }
+        }
+        writer.println()
+    }
+
+    enum class Section(val title: String, val level: Int = 1) {
```

Comment:
✏️ Probably better to extract as a top-level enum class at this point since it's shared.

Reply:
Makes sense